### PR TITLE
Feature/saas 9815 aws ensure that cognito have m

### PR DESCRIPTION
**Pull request by abdullahaslam306**

### CODE REVIEW

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+            }
+            for (let userPool of userPools.data) {
+                var describeUserPool = helpers.addSource(cache, source,
+                    ['cognitoidentityserviceprovider', 'describeUserPool', region, userPool.Id]);
+
+                if (!describeUserPool || describeUserPool.err || !describeUserPool.data){
+                    helpers.addResult(results, 3,
+                        'Unable to query for Cognito: ' + helpers.addError(describeUserPool), region);
+                    
+                    return rcb();
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoMFAEnabled.js:41]:
```suggestion
                   continue;
```

Comment (alphadev4) [reply plugins/aws/cognito/cognitoMFAEnabled.js:41]: use continue in for loop

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+            if (userPools.err || !userPools.data){
+                helpers.addResult(results, 3,  'Unable to query api: ' + helpers.addError(userPools), region);
+                return rcb();
+            }
+            if (!userPools.data.length){
+                helpers.addResult(results, 0, 'No User pool found', region);
+                return rcb();
+            }
+            for (let userPool of userPools.data) {
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoMFAEnabled.js:33]:
```suggestion
            for (let userPool of userPools.data) {
                  if (!userPool.Id) continue;
```

Comment (alphadev4) [reply plugins/aws/cognito/cognitoMFAEnabled.js:33]: add this check to make sure userPool.Id is present

File: plugins/aws/cognito/cognitoMFAEnabled.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (alphadev4) [reply plugins/aws/cognito/cognitoMFAEnabled.js:41]: use continue in for loop

File: plugins/aws/cognito/cognitoMFAEnabled.spec.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+];
+const describeUserPool = [
+    {
+      Id: 'us-east-1_cbDh8sCFGH',
+      Name: 'test',
+      LambdaConfig: {},
+      MfaConfiguration: 'OFF',
+      EstimatedNumberOfUsers: 0,
+      UserPoolTags: {},
+      Arn: 'arn:aws:cognito-idp:us-east-1:10136388154:userpool/us-east-1_cbDh8sCFGH',
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoMFAEnabled.spec.js:19]:
```suggestion
      Arn: 'arn:aws:cognito-idp:us-east-1:1111222222222:userpool/us-east-1_cbDh8sCFGH',
```

Comment (alphadev4) [reply plugins/aws/cognito/cognitoMFAEnabled.spec.js:19]: dummy id

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    run: function(cache, settings, callback) {
+        var results = [];
+        var source = {};
+        var regions = helpers.regions(settings);
+  
+        async.each(regions.cognitoidentityserviceprovider, function(region, rcb){
+            var userPools = helpers.addSource(cache, source,
+                ['cognitoidentityserviceprovider', 'listUserPools', region]);
+            
+
```
[the remark this answers is not in the document]
Comment (alphadev4) [reply plugins/aws/cognito/cognitoMFAEnabled.js:22]: -

**Review by alphadev4 — CHANGES_REQUESTED** [review]

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            if (userPools.err || !userPools.data){
+                helpers.addResult(results, 3,  'Unable to query api: ' + helpers.addError(userPools), region);
+                return rcb();
+            }
+            if (!userPools.data.length){
+                helpers.addResult(results, 0, 'No User pool found', region);
+                return rcb();
+            }
+            for (let userPool of userPools.data) {
+
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoMFAEnabled.js:33]:
```suggestion
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                if (!describeUserPool || describeUserPool.err || !describeUserPool.data){
+                    helpers.addResult(results, 3,
+                        'Unable to query for Cognito: ' + helpers.addError(describeUserPool), region);
+                    
+                    continue;
+                }
+                if (describeUserPool.data.MfaConfiguration && describeUserPool.data.MfaConfiguration == 'ON'){
+                    helpers.addResult(results, 0, 'Cognito has MFA enabled', region, describeUserPool.data.Arn);
+                } else {
+                    helpers.addResult(results, 2, 'MFA not enabled for cognito', region, describeUserPool.data.Arn);
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoMFAEnabled.js:47]:
```suggestion
                    helpers.addResult(results, 2, 'Cognito does not have MFA enabled', region, describeUserPool.data.Arn);
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            for (let userPool of userPools.data) {
+                if (!userPool.Id) continue;
+
+                var describeUserPool = helpers.addSource(cache, source,
+                    ['cognitoidentityserviceprovider', 'describeUserPool', region, userPool.Id]);
+
+                if (!describeUserPool || describeUserPool.err || !describeUserPool.data){
+                    helpers.addResult(results, 3,
+                        'Unable to query for Cognito: ' + helpers.addError(describeUserPool), region);
+
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoMFAEnabled.js:41]:
```suggestion
```

**Review by alphadev4 — APPROVED** [review]

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Cognito User Pool Has MFA enabled',
+    category: 'CognitoIdentityServiceProvider',
+    domain: 'Identity Service Provider',
+    description: 'Ensure that cognito user pool has MFA enabled.',
+    more_info: 'Enabling WAF allows control over requests to the load balancer, allowing or denying traffic based off rules in the Web ACL',
+    link: 'https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-mfa.html',
+    recommended_action: '1. Enter the Cognito service. 2. Enter user pools and enable MFA from signin experience ',
```
Comment (alphadev4) [inline plugins/aws/cognito/cognitoMFAEnabled.js:11]:
```suggestion
    recommended_action: '1. Enter the Cognito service. 2. Enter user pools and enable MFA from sign in experience',
```

**Review by alphadev4 — APPROVED** [review]

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Cognito User Pool Has MFA enabled',
+    category: 'CognitoIdentityServiceProvider',
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:6]:
```suggestion
    title: 'Cognito User Pool MFA enabled',
    category: 'Cognito',
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+    domain: 'Identity Service Provider',
+    description: 'Ensure that cognito user pool has MFA enabled.',
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:8]:
```suggestion
    description: 'Ensure that Cognito user pool has MFA enabled.',
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    more_info: 'Enabling WAF allows control over requests to the load balancer, allowing or denying traffic based off rules in the Web ACL',
+    link: 'https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-settings-mfa.html',
+    recommended_action: '1. Enter the Cognito service. 2. Enter user pools and enable MFA from sign in experience',
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:11]:
```suggestion
    recommended_action: '1. Enter the Cognito service. 2. Enter user pools and enable MFA from sign in experience.',
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        var regions = helpers.regions(settings);
+  
+        async.each(regions.cognitoidentityserviceprovider, function(region, rcb){
+            var userPools = helpers.addSource(cache, source,
+                ['cognitoidentityserviceprovider', 'listUserPools', region]);
+            
+            if (!userPools) return rcb();
+
+            if (userPools.err || !userPools.data){
+                helpers.addResult(results, 3,  'Unable to query api: ' + helpers.addError(userPools), region);
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:25]:
```suggestion
                helpers.addResult(results, 3,  'Unable to query Cognito user pools: ' + helpers.addError(userPools), region);
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+                return rcb();
+            }
+            if (!userPools.data.length){
+                helpers.addResult(results, 0, 'No User pool found', region);
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:29]:
```suggestion
                helpers.addResult(results, 0, 'No Cognito user pools found', region);
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+                return rcb();
+            }
+            for (let userPool of userPools.data) {
+                if (!userPool.Id) continue;
+
+                var describeUserPool = helpers.addSource(cache, source,
+                    ['cognitoidentityserviceprovider', 'describeUserPool', region, userPool.Id]);
+                if (!describeUserPool || describeUserPool.err || !describeUserPool.data || !describeUserPool.data.UserPool){
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:37]:
```suggestion
                    ['cognitoidentityserviceprovider', 'describeUserPool', region, userPool.Id]);

                if (!describeUserPool || describeUserPool.err || !describeUserPool.data || !describeUserPool.data.UserPool){
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+            if (userPools.err || !userPools.data){
+                helpers.addResult(results, 3,  'Unable to query Cognito user pools: ' + helpers.addError(userPools), region);
+                return rcb();
+            }
+            if (!userPools.data.length){
+                helpers.addResult(results, 0, 'No Cognito user pools found', region);
+                return rcb();
+            }
+            for (let userPool of userPools.data) {
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:32]:
```suggestion
            }

            if (!userPools.data.length){
                helpers.addResult(results, 0, 'No Cognito user pools found', region);
                return rcb();
            }

            for (let userPool of userPools.data) {
```

File: plugins/aws/cognito/cognitoMFAEnabled.spec.js [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+};
+
+describe('cognitoMFAEnabled', function () {
+    describe('run', function () {
+        it('should give unknown result if unable to list user pools', function (done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(3);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Unable to query api:');
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.spec.js:61]:
```suggestion
                expect(results[0].message).to.include('Unable to query Cognito user pools:');
```

File: plugins/aws/cognito/cognitoMFAEnabled.spec.js [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+            const cache = createCache(null, []);
+            cognitoMFAEnabled.run(cache, {}, callback);
+        });
+
+        it('should give passing result if User pool not found.', function (done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('No User pool found');
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.spec.js:74]:
```suggestion
                expect(results[0].message).to.include('No Cognito user pool found');
```

File: plugins/aws/cognito/cognitoMFAEnabled.spec.js [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+            const cache = createCache([], null);
+            cognitoMFAEnabled.run(cache, {}, callback);
+        });
+
+        it('should give unknown result if unable to query describe user pool api', function (done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(3);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Unable to query for Cognito');
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.spec.js:86]:
```suggestion
                expect(results[0].message).to.include('Unable to describe Cognito user pool');
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+                    ['cognitoidentityserviceprovider', 'describeUserPool', region, userPool.Id]);
+
+                if (!describeUserPool || describeUserPool.err || !describeUserPool.data || !describeUserPool.data.UserPool){
+                    helpers.addResult(results, 3,
+                        'Unable to describe Cognito user pool: ' + helpers.addError(describeUserPool), region);
+                    continue;
+                }
+
+                if (describeUserPool.data.UserPool.MfaConfiguration && describeUserPool.data.UserPool.MfaConfiguration.toUpperCase() == 'ON'){
+                    helpers.addResult(results, 0, 'Cognito has MFA enabled', region, describeUserPool.data.Arn);
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:47]:
```suggestion
                    helpers.addResult(results, 0, 'User pool has MFA enabled', region, describeUserPool.data.Arn);
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+                } else {
+                    helpers.addResult(results, 2, 'Cognito does not have MFA enabled', region, describeUserPool.data.Arn);
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:49]:
```suggestion
                    helpers.addResult(results, 2, 'User pool does not have MFA enabled', region, describeUserPool.data.Arn);
```

File: plugins/aws/cognito/cognitoMFAEnabled.spec.js [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+            const cache = createCache([listUserPools[0]], null);
+            cognitoMFAEnabled.run(cache, {}, callback);
+        });
+
+        it('should give passing result if User pool has MFA enabled', function (done) {
+            const callback = (err, results) => {
+                expect(results.length).to.equal(1);
+                expect(results[0].status).to.equal(0);
+                expect(results[0].region).to.equal('us-east-1');
+                expect(results[0].message).to.include('Cognito has MFA enabled');
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.spec.js:99]:
```suggestion
                expect(results[0].message).to.include('User pool has MFA enabled');
```

File: plugins/aws/cognito/cognitoMFAEnabled.spec.js [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+            const cache = createCache([listUserPools[0]], describeUserPool[1]);
+            cognitoMFAEnabled.run(cache, {}, callback);
+        });
+
+        it('should give failing result if User pool does not have MFA enabled', function (done) {
+                const callback = (err, results) => {
+                    expect(results.length).to.equal(1);
+                    expect(results[0].status).to.equal(2);
+                    expect(results[0].region).to.equal('us-east-1');
+                    expect(results[0].message).to.include('Cognito does not have MFA enabled');
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.spec.js:112]:
```suggestion
                    expect(results[0].message).to.include('User pool does not have MFA enabled');
```

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+
+            for (let userPool of userPools.data) {
+                if (!userPool.Id) continue;
+
+                var describeUserPool = helpers.addSource(cache, source,
+                    ['cognitoidentityserviceprovider', 'describeUserPool', region, userPool.Id]);
+
+                if (!describeUserPool || describeUserPool.err || !describeUserPool.data || !describeUserPool.data.UserPool){
+                    helpers.addResult(results, 3,
+                        'Unable to describe Cognito user pool: ' + helpers.addError(describeUserPool), region);
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:42]: Where is resource? You need to create ARN above. Also use that ARN for messages below

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

File: helpers/aws/api_multipart.js [hunk 24]
```diff
@@ -876,6 +876,7 @@ var calls = [
                 paginate: 'PaginationToken'
             }
         },
+
```
Comment (AkhtarAmir) [inline helpers/aws/api_multipart.js:879]:
```suggestion
```

File: helpers/aws/regions.js [hunk 25]
```diff
@@ -47,6 +47,11 @@ module.exports = {
     cloudformation: [...regions, ...newRegions],
     cloudtrail: [...regions, ...newRegions],
     cloudwatchlogs: [...regions, ...newRegions],
+    cognitoidentityserviceprovider: [ 'us-east-1', 'us-east-2', 'us-west-1', 'us-west-2', 'ca-central-1','eu-central-1',     
+        'eu-west-1','eu-west-2', 'eu-west-3','eu-north-1', 'eu-south-1','ap-northeast-1','ap-northeast-2',
+        'ap-southeast-1','ap-northeast-3', 'ap-south-1', 'sa-east-1', 'me-south-1' 
+    ],
+
```
Comment (AkhtarAmir) [inline helpers/aws/regions.js:54]:
```suggestion
```

**Review by AkhtarAmir — APPROVED** [review]

File: plugins/aws/cognito/cognitoMFAEnabled.js [hunk 26]
```diff
@@ -0,0 +1,64 @@
+var async = require('async');
+var helpers = require('../../../helpers/aws');
+
+module.exports = {
+    title: 'Cognito User Pool MFA enabled',
+    category: 'Cognito',
+    domain: 'Identity Service Provider',
```
Comment (AkhtarAmir) [inline plugins/aws/cognito/cognitoMFAEnabled.js:7]:
```suggestion
    domain: 'Identity and Access Management',
```

**Review by AkhtarAmir — APPROVED** [review]

**Pull request merged**